File: frontend/src/models/Post.ts
import User, { mapUser } from "./User";

export default interface Post {
  id: number;
  author: User;
  title: string;
  content: string;
  createdAt: Date;
  updatedAt?: Date;
}

export function mapPost(data: any): Post {
  return {
    id: data.id,
    author: mapUser(data.author),
    title: data.title,
    content: data.content,
    createdAt: new Date(Date.parse(data.created_at)),
    updatedAt: data.updated_at && new Date(Date.parse(data.updated_at)),
  }
}

File: backend/src/models/post.rs
use rocket::serde::{Serialize, Deserialize};
use chrono::{DateTime, Utc};
use validator::{Validate};

use crate::models::user::{User};

#[derive(Debug, Serialize)]
pub struct Post {
  pub id: i64,
  pub title: String,
  pub content: String,
  pub created_at: DateTime<Utc>,
  pub updated_at: Option<DateTime<Utc>>,
  pub author: User,
}

#[derive(Debug, Validate, Deserialize)]
pub struct NewPost {
  #[validate(length(min=1, message=r#"must not be empty."#))]
  pub title: String,
  #[validate(length(min=1, message=r#"must not be empty."#))]
  pub content: String,
}

#[derive(Debug, Validate, Deserialize)]
pub struct UpdatePost {
  #[validate(length(min=1, message=r#"must not be empty."#))]
  pub title: String,
  #[validate(length(min=1, message=r#"must not be empty."#))]
  pub content: String,
}
